Machine ID 28
PRN 230340125007

In [43]:
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error, ConfusionMatrixDisplay,  confusion_matrix, accuracy_score
from sklearn.model_selection import cross_val_score, GridSearchCV, KFold
from xgboost import XGBRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.preprocessing import StandardScaler, MinMaxScaler 
from matplotlib import pyplot as plt
from sklearn.pipeline import Pipeline
import os


metric used for the comparison of algorithms is mean_squared_error since it is regression problem

In [24]:
os.chdir(r"C:\Users\Administrator.DBDA\Desktop\PML_SET_A")

In [25]:
house = pd.read_csv("california.csv")
house

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

In [26]:
X = house.drop(['MedHouseVal'], axis = 1)
y = house['MedHouseVal']

## XGBRegressor

In [36]:
clf = XGBRegressor(random_state=23)
params = {'learning_rate':[0.001, 0.002],
          'max_depth':[2,3,4,5,6,None],
          'n_estimators':[50,100,150]}
kfold = KFold(n_splits=5, shuffle=True, random_state=23)
cv = GridSearchCV(clf, param_grid=params,cv=kfold, verbose=3, scoring='neg_mean_squared_error')

cv.fit(X,y)
df_cv = pd.DataFrame(cv.cv_results_)

print(cv.best_params_)
print(cv.best_score_)

Fitting 5 folds for each of 36 candidates, totalling 180 fits
[CV 1/5] END learning_rate=0.001, max_depth=2, n_estimators=50;, score=-3.519 total time=   0.0s
[CV 2/5] END learning_rate=0.001, max_depth=2, n_estimators=50;, score=-3.537 total time=   0.0s
[CV 3/5] END learning_rate=0.001, max_depth=2, n_estimators=50;, score=-3.488 total time=   0.0s
[CV 4/5] END learning_rate=0.001, max_depth=2, n_estimators=50;, score=-3.412 total time=   0.0s
[CV 5/5] END learning_rate=0.001, max_depth=2, n_estimators=50;, score=-3.550 total time=   0.0s
[CV 1/5] END learning_rate=0.001, max_depth=2, n_estimators=100;, score=-3.253 total time=   0.1s
[CV 2/5] END learning_rate=0.001, max_depth=2, n_estimators=100;, score=-3.273 total time=   0.2s
[CV 3/5] END learning_rate=0.001, max_depth=2, n_estimators=100;, score=-3.228 total time=   0.1s
[CV 4/5] END learning_rate=0.001, max_depth=2, n_estimators=100;, score=-3.152 total time=   0.2s
[CV 5/5] END learning_rate=0.001, max_depth=2, n_estimators=1

[CV 5/5] END learning_rate=0.001, max_depth=None, n_estimators=100;, score=-3.237 total time=   0.6s
[CV 1/5] END learning_rate=0.001, max_depth=None, n_estimators=150;, score=-2.932 total time=   1.0s
[CV 2/5] END learning_rate=0.001, max_depth=None, n_estimators=150;, score=-2.960 total time=   1.0s
[CV 3/5] END learning_rate=0.001, max_depth=None, n_estimators=150;, score=-2.924 total time=   1.0s
[CV 4/5] END learning_rate=0.001, max_depth=None, n_estimators=150;, score=-2.849 total time=   1.0s
[CV 5/5] END learning_rate=0.001, max_depth=None, n_estimators=150;, score=-2.975 total time=   1.0s
[CV 1/5] END learning_rate=0.002, max_depth=2, n_estimators=50;, score=-3.252 total time=   0.0s
[CV 2/5] END learning_rate=0.002, max_depth=2, n_estimators=50;, score=-3.273 total time=   0.0s
[CV 3/5] END learning_rate=0.002, max_depth=2, n_estimators=50;, score=-3.228 total time=   0.0s
[CV 4/5] END learning_rate=0.002, max_depth=2, n_estimators=50;, score=-3.152 total time=   0.0s
[CV 5/

[CV 4/5] END learning_rate=0.002, max_depth=None, n_estimators=50;, score=-3.105 total time=   0.3s
[CV 5/5] END learning_rate=0.002, max_depth=None, n_estimators=50;, score=-3.236 total time=   0.3s
[CV 1/5] END learning_rate=0.002, max_depth=None, n_estimators=100;, score=-2.692 total time=   0.6s
[CV 2/5] END learning_rate=0.002, max_depth=None, n_estimators=100;, score=-2.721 total time=   0.6s
[CV 3/5] END learning_rate=0.002, max_depth=None, n_estimators=100;, score=-2.691 total time=   0.6s
[CV 4/5] END learning_rate=0.002, max_depth=None, n_estimators=100;, score=-2.616 total time=   0.6s
[CV 5/5] END learning_rate=0.002, max_depth=None, n_estimators=100;, score=-2.738 total time=   0.6s
[CV 1/5] END learning_rate=0.002, max_depth=None, n_estimators=150;, score=-2.277 total time=   1.0s
[CV 2/5] END learning_rate=0.002, max_depth=None, n_estimators=150;, score=-2.309 total time=   1.0s
[CV 3/5] END learning_rate=0.002, max_depth=None, n_estimators=150;, score=-2.287 total time=

conclusion: the mean squared erorr for xgb regressor is 2.28

## KNN 

In [44]:
scaler = StandardScaler()
knn = KNeighborsRegressor()
params = {'KNN__n_neighbors': np.arange(1,31)}
pipe = Pipeline([('SCL',scaler),('KNN',knn)])
kfold = KFold(n_splits=5, shuffle=True, random_state=23)
gcv = GridSearchCV(pipe, param_grid=params,verbose=3,
                   scoring='neg_mean_squared_error',
                   cv=kfold)
gcv.fit(X, y)
print(gcv.best_params_)
print(gcv.best_score_)

Fitting 5 folds for each of 30 candidates, totalling 150 fits
[CV 1/5] END ...............KNN__n_neighbors=1;, score=-0.587 total time=   0.2s
[CV 2/5] END ...............KNN__n_neighbors=1;, score=-0.655 total time=   0.1s
[CV 3/5] END ...............KNN__n_neighbors=1;, score=-0.683 total time=   0.1s
[CV 4/5] END ...............KNN__n_neighbors=1;, score=-0.611 total time=   0.1s
[CV 5/5] END ...............KNN__n_neighbors=1;, score=-0.654 total time=   0.1s
[CV 1/5] END ...............KNN__n_neighbors=2;, score=-0.444 total time=   0.1s
[CV 2/5] END ...............KNN__n_neighbors=2;, score=-0.492 total time=   0.2s
[CV 3/5] END ...............KNN__n_neighbors=2;, score=-0.517 total time=   0.2s
[CV 4/5] END ...............KNN__n_neighbors=2;, score=-0.469 total time=   0.2s
[CV 5/5] END ...............KNN__n_neighbors=2;, score=-0.527 total time=   0.1s
[CV 1/5] END ...............KNN__n_neighbors=3;, score=-0.398 total time=   0.2s
[CV 2/5] END ...............KNN__n_neighbors=3;

[CV 2/5] END ..............KNN__n_neighbors=21;, score=-0.407 total time=   0.3s
[CV 3/5] END ..............KNN__n_neighbors=21;, score=-0.433 total time=   0.3s
[CV 4/5] END ..............KNN__n_neighbors=21;, score=-0.405 total time=   0.3s
[CV 5/5] END ..............KNN__n_neighbors=21;, score=-0.434 total time=   0.3s
[CV 1/5] END ..............KNN__n_neighbors=22;, score=-0.362 total time=   0.4s
[CV 2/5] END ..............KNN__n_neighbors=22;, score=-0.408 total time=   0.3s
[CV 3/5] END ..............KNN__n_neighbors=22;, score=-0.433 total time=   0.3s
[CV 4/5] END ..............KNN__n_neighbors=22;, score=-0.406 total time=   0.3s
[CV 5/5] END ..............KNN__n_neighbors=22;, score=-0.434 total time=   0.3s
[CV 1/5] END ..............KNN__n_neighbors=23;, score=-0.363 total time=   0.3s
[CV 2/5] END ..............KNN__n_neighbors=23;, score=-0.409 total time=   0.3s
[CV 3/5] END ..............KNN__n_neighbors=23;, score=-0.435 total time=   0.3s
[CV 4/5] END ..............K

conclusion: the mean squared error for knn regressor is 0.39

## Random Forest

In [48]:
rf = RandomForestRegressor(random_state=23)
kfold = KFold(n_splits=5,random_state=23,shuffle=True)
rf.get_params()

params = {'max_features': np.arange(2,6)}
gcv = GridSearchCV(rf, param_grid=params , verbose=3,cv=kfold, scoring="neg_mean_squared_error")
gcv.fit(X,y)
gcv.best_score_
gcv.best_params_
print(gcv.best_score_)
print(gcv.best_params_)

Fitting 5 folds for each of 4 candidates, totalling 20 fits
[CV 1/5] END ...................max_features=2;, score=-0.244 total time=  10.9s
[CV 2/5] END ...................max_features=2;, score=-0.227 total time=  10.9s
[CV 3/5] END ...................max_features=2;, score=-0.256 total time=  11.2s
[CV 4/5] END ...................max_features=2;, score=-0.235 total time=  10.9s
[CV 5/5] END ...................max_features=2;, score=-0.257 total time=  11.1s
[CV 1/5] END ...................max_features=3;, score=-0.238 total time=  15.3s
[CV 2/5] END ...................max_features=3;, score=-0.224 total time=  15.5s
[CV 3/5] END ...................max_features=3;, score=-0.255 total time=  15.4s
[CV 4/5] END ...................max_features=3;, score=-0.228 total time=  15.3s
[CV 5/5] END ...................max_features=3;, score=-0.252 total time=  15.4s
[CV 1/5] END ...................max_features=4;, score=-0.243 total time=  19.6s
[CV 2/5] END ...................max_features=4;, 

conclusion the mean squared error for the random forest is 0.23 which is best as compared to knn and xgb regressor.
indicating that it has the smallest discrepancy between the model's predictions and the true observations